nomos-core/chain-defs/src/sdp/state.rs

```diff
@@ -127,101 +127,82 @@ pub enum ProviderStateError {
     BlockFromPast,
 }
 
-#[derive(Debug, Eq, PartialEq, Hash, Clone)]
-pub enum TransientDeclarationState {
-    Active(ActiveState),
-    Inactive(InactiveState),
-    Withdrawn(WithdrawnState),
+#[derive(Debug, Eq, PartialEq)]
+pub enum TransientDeclarationState<'a> {
+    Active(ActiveState<'a>),
+    Inactive(InactiveState<'a>),
+    Withdrawn(WithdrawnState<'a>),
 }
 
-impl From<TransientDeclarationState> for DeclarationInfo {
-    fn from(state: TransientDeclarationState) -> Self {
-        match state {
-            TransientDeclarationState::Active(active_state) => active_state.0,
-            TransientDeclarationState::Inactive(inactive_state) => inactive_state.0,
-            TransientDeclarationState::Withdrawn(withdrawn_state) => withdrawn_state.0,
-        }
-    }
-}
-
-impl TransientDeclarationState {
-    pub fn try_from_info(
+impl<'a> TransientDeclarationState<'a> {
+    pub fn try_from_state(
         block_number: BlockNumber,
-        declaration_info: DeclarationInfo,
+        declaration_state: &'a mut DeclarationState,
         service_params: &ServiceParameters,
-    ) -> Result<Self, ProviderStateError> {
-        if declaration_info.created > block_number {
-            return Err(ProviderStateError::BlockFromPast);
+    ) -> Result<Self, DeclarationStateError> {
+        if declaration_state.created > block_number {
+            return Err(DeclarationStateError::BlockFromPast);
         }
 
-        if declaration_info.withdrawn.is_some() {
-            return Ok(WithdrawnState(declaration_info).into());
+        if declaration_state.withdrawn.is_some() {
+            return Ok(WithdrawnState(declaration_state).into());
         }
 
         // This section checks if recently created provider is still considered active
         // even without having activity recorded yet.
-        if declaration_info
+        if declaration_state
             .created
             .wrapping_add(service_params.inactivity_period)
             > block_number
         {
-            return Ok(ActiveState(declaration_info).into());
+            return Ok(ActiveState(declaration_state).into());
         }
 
         // Check if provider has ever got the activity recorded first and then see if
         // the activity record was recent.
-        if let Some(activity) = declaration_info.active {
+        if let Some(activity) = declaration_state.active {
```

Comment:
you can probably merge this with the check above, either by setting activity at creation as in the spec or by doing the check on `max(activity, created)`

Reply:
> you can probably merge this with the check above

Will do.